oshi-core/src/main/java/oshi/software/os/unix/freebsd/FreeBsdOSThread.java

```diff
@@ -0,0 +1,162 @@
+    }
+
+    @Override
+    public OSProcess.State getState() {
+        return this.state;
+    }
+
+    @Override
+    public long getStartMemoryAddress() {
+        return this.startMemoryAddress;
+    }
+
+    @Override
+    public long getContextSwitches() {
+        return this.contextSwitches;
+    }
+
+    @Override
+    public long getMinorFaults() {
+        return this.minorFaults;
+    }
+
+    @Override
+    public long getMajorFaults() {
+        return this.majorFaults;
+    }
+
+    @Override
+    public long getKernelTime() {
+        return this.kernelTime;
+    }
+
+    @Override
+    public long getUserTime() {
+        return this.userTime;
+    }
+
+    @Override
+    public long getUpTime() {
+        return this.upTime;
+    }
+
+    @Override
+    public long getStartTime() {
+        return this.startTime;
+    }
+
+    @Override
+    public boolean updateAttributes() {
+        String psCommand = "ps -awwxo tdname,lwp,state,etimes,systime,time,tdaddr,nivcsw,nvcsw,majflt -H -p "
+                + getOwningProcessId() + " | grep " + getThreadId(); //there is no filter for thread is ps command, hence using grep
+        List<String> threadList = ExecutingCommand.runNative(psCommand);
+        if (threadList.size() > 1) {
+            // skip header row
+            String[] split = ParseUtil.whitespaces.split(threadList.get(1).trim(), 16);
+            if (split.length == 10) {
+                return updateAttributes(split);
+            }
+        }
+        this.state = OSProcess.State.INVALID;
+        return false;
+    }
+
+    private boolean updateAttributes(String[] split) {
+        this.name = split[0];
+        this.threadId = ParseUtil.parseIntOrDefault(split[1], 0);
+        switch (split[2].charAt(0)) {
+            case 'R':
+                this.state = OSProcess.State.RUNNING;
+                break;
+            case 'I':
+            case 'S':
+                this.state = OSProcess.State.SLEEPING;
+                break;
+            case 'D':
+            case 'L':
+            case 'U':
+                this.state = OSProcess.State.WAITING;
+                break;
+            case 'Z':
+                this.state = OSProcess.State.ZOMBIE;
+                break;
+            case 'T':
+                this.state = OSProcess.State.STOPPED;
+                break;
+            default:
+                this.state = OSProcess.State.OTHER;
+                break;
+        }
+        // Avoid divide by zero for processes up less than a second
+        long elapsedTime = ParseUtil.parseDHMSOrDefault(split[3], 0L); //etimes
+        this.upTime = elapsedTime < 1L ? 1L : elapsedTime;
+        long now = System.currentTimeMillis();
+        this.startTime = now - this.upTime;
+        this.kernelTime = ParseUtil.parseDHMSOrDefault(split[4], 0L); //systime
+        this.userTime = ParseUtil.parseDHMSOrDefault(split[5], 0L) - this.kernelTime; //time
+        //this.startMemoryAddress = split[6]; hexadecimal
```

Comment:
Should we change the memory address data type to string, as freebsd is returning hexdecimal value?

Reply:
Parsing hex is easy using `Long.decode()`.  See `ParseUtil.parseLastLong()`.

Reply:
Actually since the string doesn't have the `0x` it's probably better to just use `Long.parseLong(split[6], 16)`.  (Except don't do it here, put it in the `ParseUtil` class and catch the `NumberFormatException` and return a default if it fails.)

Reply:
ok.